stan/math/prim/arr/fun/array_builder.hpp

```diff
@@ -0,0 +1,36 @@
+#ifndef STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+#define STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+
+#include <stan/math/prim/arr/fun/promoter.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Structure for building up arrays in an expression (rather than
+     * in statements) using an argumentchaining add() method and
+     * a getter method array() to return the result.
+     *
+     * @tparam T1 type of arg1
+     */
+    template <typename T>
+    struct array_builder {
+      std::vector<T> x_;
+      array_builder() : x_() { }
+      template <typename F>
+      array_builder& add(const F& u) {
+        T t;
+        promoter<F, T>::promote(u, t);
+        x_.push_back(t);
+        return *this;
+      }
+
+      std::vector<T> array() {
```

Comment:
The return can probably be `const std::vector<T>&` --- constant reference.  It'd mean the result only lasts as long as the array builder, but the client could also copy it out if necessary.